### PR TITLE
Ignore non-verack and non-version messages in handshake

### LINKED ISSUE #3429: ZG-CONFORMANCE-003-006: ignore non-Version/Verack messages during the handshake
## Motivation

The Ziggurat team picked up some handshake bugs we should fix:
https://github.com/eqlabs/ziggurat#conformance
https://github.com/eqlabs/ziggurat/blob/master/SPEC.md#zg-conformance-003

## Tasks

In the handshake:
- ignore unexpected messages while waiting for `Version`
- ignore unexpected messages while waiting for `Verack`

### PR DESCRIPTION
**Pull request by oxarbitrage**
## Motivation

We want to ignore some messages during handshake to make the conformance tests of ziggurat pass: https://github.com/eqlabs/ziggurat/blob/master/SPEC.md#zg-conformance-003

Close https://github.com/ZcashFoundation/zebra/issues/3429

## Solution

Ignore the verack can be done just by not closing the connection when receiving other messages at the end of the handshake. This is done in https://github.com/ZcashFoundation/zebra/commit/cce709ab8509fa94ebf8950b477cf6ecba4aa91c

The version is a bit trickier, we need to wait for another message when we receive a non version one at the beginning of the handshake.

I made a proof of concept of this at https://github.com/ZcashFoundation/zebra/commit/cce709ab8509fa94ebf8950b477cf6ecba4aa91c however the code can be improved as it is doing some duplicated work. I am looking for some suggestions there but before that i want to make sure that we will want to really do that as it is in this pull request.

## Review

I will like @teor2345 and/or @jvff to take a look.

### Reviewer Checklist

  - [ ] Code implements Specs and Designs
  - [ ] Tests for Expected Behaviour
  - [ ] Tests for Errors

### CODE REVIEW

**Comment by oxarbitrage** [comment]
The ziggurat tests are a bit intermittent from here, sometimes stuff timeout, however with the code of this pull request the following tests that were always failing before now pass:

Verack:
```
$ cargo test ignore_message_inplace_of_verack -- --test-threads=1 
    Finished test [unoptimized + debuginfo] target(s) in 0.02s
     Running unittests (target/debug/deps/ziggurat-e77efbb62339a887)

running 22 tests
test tests::conformance::handshake::ignore_message_inplace_of_verack::when_node_initiates_connection::addr ... ok
test tests::conformance::handshake::ignore_message_inplace_of_verack::when_node_initiates_connection::get_addr ... ok
test tests::conformance::handshake::ignore_message_inplace_of_verack::when_node_initiates_connection::get_blocks ... ok
test tests::conformance::handshake::ignore_message_inplace_of_verack::when_node_initiates_connection::get_data_block ... ok
test tests::conformance::handshake::ignore_message_inplace_of_verack::when_node_initiates_connection::get_data_tx ... ok
test tests::conformance::handshake::ignore_message_inplace_of_verack::when_node_initiates_connection::get_headers ... ok
test tests::conformance::handshake::ignore_message_inplace_of_verack::when_node_initiates_connection::inv ... ok
test tests::conformance::handshake::ignore_message_inplace_of_verack::when_node_initiates_connection::mempool ... ok
test tests::conformance::handshake::ignore_message_inplace_of_verack::when_node_initiates_connection::not_found ... ok
test tests::conformance::handshake::ignore_message_inplace_of_verack::when_node_initiates_connection::ping ... ok
test tests::conformance::handshake::ignore_message_inplace_of_verack::when_node_initiates_connection::pong ... ok
test tests::conformance::handshake::ignore_message_inplace_of_verack::when_node_receives_connection::addr ... ok
test tests::conformance::handshake::ignore_message_inplace_of_verack::when_node_receives_connection::get_addr ... ok
test tests::conformance::handshake::ignore_message_inplace_of_verack::when_node_receives_connection::get_blocks ... ok
test tests::conformance::handshake::ignore_message_inplace_of_verack::when_node_receives_connection::get_data_block ... ok
test tests::conformance::handshake::ignore_message_inplace_of_verack::when_node_receives_connection::get_data_tx ... ok
test tests::conformance::handshake::ignore_message_inplace_of_verack::when_node_receives_connection::get_headers ... ok
test tests::conformance::handshake::ignore_message_inplace_of_verack::when_node_receives_connection::inv ... ok
test tests::conformance::handshake::ignore_message_inplace_of_verack::when_node_receives_connection::mempool ... ok
test tests::conformance::handshake::ignore_message_inplace_of_verack::when_node_receives_connection::not_found ... ok
test tests::conformance::handshake::ignore_message_inplace_of_verack::when_node_receives_connection::ping ... ok
test tests::conformance::handshake::ignore_message_inplace_of_verack::when_node_receives_connection::pong ... ok
```

Version
```
$ cargo test ignore_message_inplace_of_version -- --test-threads=1 
    Finished test [unoptimized + debuginfo] target(s) in 0.02s
     Running unittests (target/debug/deps/ziggurat-e77efbb62339a887)

running 24 tests
test tests::conformance::handshake::ignore_message_inplace_of_version::when_node_initiates_connection::addr ... ok
test tests::conformance::handshake::ignore_message_inplace_of_version::when_node_initiates_connection::get_addr ... ok
test tests::conformance::handshake::ignore_message_inplace_of_version::when_node_initiates_connection::get_blocks ... ok
test tests::conformance::handshake::ignore_message_inplace_of_version::when_node_initiates_connection::get_data_block ... ok
test tests::conformance::handshake::ignore_message_inplace_of_version::when_node_initiates_connection::get_data_tx ... ok
test tests::conformance::handshake::ignore_message_inplace_of_version::when_node_initiates_connection::get_headers ... ok
test tests::conformance::handshake::ignore_message_inplace_of_version::when_node_initiates_connection::inv ... ok
test tests::conformance::handshake::ignore_message_inplace_of_version::when_node_initiates_connection::mempool ... ok
test tests::conformance::handshake::ignore_message_inplace_of_version::when_node_initiates_connection::not_found ... ok
test tests::conformance::handshake::ignore_message_inplace_of_version::when_node_initiates_connection::ping ... ok
test tests::conformance::handshake::ignore_message_inplace_of_version::when_node_initiates_connection::pong ... ok
test tests::conformance::handshake::ignore_message_inplace_of_version::when_node_initiates_connection::verack ... ok
test tests::conformance::handshake::ignore_message_inplace_of_version::when_node_receives_connection::addr ... ok
test tests::conformance::handshake::ignore_message_inplace_of_version::when_node_receives_connection::get_addr ... ok
test tests::conformance::handshake::ignore_message_inplace_of_version::when_node_receives_connection::get_blocks ... ok
test tests::conformance::handshake::ignore_message_inplace_of_version::when_node_receives_connection::get_data_block ... ok
test tests::conformance::handshake::ignore_message_inplace_of_version::when_node_receives_connection::get_data_tx ... ok
test tests::conformance::handshake::ignore_message_inplace_of_version::when_node_receives_connection::get_headers ... ok
test tests::conformance::handshake::ignore_message_inplace_of_version::when_node_receives_connection::inv ... ok
test tests::conformance::handshake::ignore_message_inplace_of_version::when_node_receives_connection::mempool ... ok
test tests::conformance::handshake::ignore_message_inplace_of_version::when_node_receives_connection::not_found ... ok
test tests::conformance::handshake::ignore_message_inplace_of_version::when_node_receives_connection::ping ... ok
test tests::conformance::handshake::ignore_message_inplace_of_version::when_node_receives_connection::pong ... ok
test tests::conformance::handshake::ignore_message_inplace_of_version::when_node_receives_connection::verack ... ok

```

**Comment by oxarbitrage** [comment]
By the way, it seems ziggurat uses the testnet protocol version so in order to make the tests run properly, the constant need to be updated. I sent a ticket to make this a bit easier as i had a hard time figuring it out by myself.

https://github.com/eqlabs/ziggurat/issues/110

File: zebra-network/src/peer/handshake.rs [hunk 1]
```diff
@@ -588,13 +588,25 @@ where
...
+    match remote_msg {
+        Message::Version { .. } => (),
+        _ => {
+            remote_msg = peer_conn
+                .next()
+                .await
+                .ok_or(HandshakeError::ConnectionClosed)??;
+            debug!(?remote_msg, "got message from remote peer");
+        }
+    }
```
Comment (teor2345) [inline zebra-network/src/peer/handshake.rs]: Blocker: This is good, but it's not quite enough.

We need to ignore any number of extra non-Version messages, then continue with the handshake as soon as we get a Version message.

(Currently, this code only ignores 1 extra non-Version message.)

File: zebra-network/src/peer/handshake.rs [hunk 2]
```diff
@@ -705,9 +717,9 @@ where
         .await
         .ok_or(HandshakeError::ConnectionClosed)??;
     if let Message::Verack = remote_msg {
-        debug!("got verack from remote peer");
+        debug!("got verack message from remote peer");
     } else {
-        Err(HandshakeError::UnexpectedMessage(Box::new(remote_msg)))?;
+        debug!("ignoring non-verack message from remote peer");
```
Comment (teor2345) [inline zebra-network/src/peer/handshake.rs]: Blocker: This passes one specific test, but it isn't quite the behaviour we want.

We need to ignore any number of non-Verack messages, then return the successful Connection as soon as we get a Verack message.

(Currently, this code continues as soon as it gets any message - it doesn't wait for a Verack message.)

The code that waits for Version messages and the code that waits for Verack messages should look very similar.

**Review by teor2345 — COMMENTED** [review]
Try using a loop to ignore messages that aren't the ones we're looking for.

File: zebra-network/src/peer/handshake.rs [hunk 3]
```diff
@@ -588,13 +588,29 @@ where
...
+            Message::Version { .. } => {
+                debug!(?remote_msg, "got version message from remote peer");
+                break;
+            }
+            _ => {
+                remote_msg = peer_conn
+                    .next()
+                    .await
+                    .ok_or(HandshakeError::ConnectionClosed)??;
+                debug!("ignoring non-version message from remote peer");
```
Comment (teor2345) [inline zebra-network/src/peer/handshake.rs]: Minor Logging: Can we log the message here, so we can see it when we are debugging connections?

```suggestion
                debug!(?remote_msg, "ignoring non-version message from remote peer");
```

File: zebra-network/src/peer/handshake.rs [hunk 4]
```diff
@@ -700,14 +716,26 @@ where
...
+            Message::Verack => {
+                debug!(?remote_msg, "got verack message from remote peer");
+                break;
+            }
+            _ => {
+                remote_msg = peer_conn
+                    .next()
+                    .await
+                    .ok_or(HandshakeError::ConnectionClosed)??;
+                debug!("ignoring non-verack message from remote peer");
```
Comment (teor2345) [inline zebra-network/src/peer/handshake.rs]:
```suggestion
                debug!(?remote_msg, "ignoring non-verack message from remote peer");
```

**Review by teor2345 — DISMISSED** [review]
This is a nice compact fix for this issue, thanks!

I just want to tweak the debugging a bit. So if peer connections start behaving differently after this change, we'll be able to see if the skipped messages.

**Review by teor2345 — APPROVED** [review]
Let's go!

**Comment by teor2345** [comment]
This change might fix some block explorers, so we should check them again when we have time:
- #1904

**Pull request merged**